src/charm.py

```diff
@@ -358,6 +434,64 @@ def _set_secret(self, scope: str, key: str, value: Optional[str]) -> None:
         else:
             raise RuntimeError("Unknown secret scope.")
 
+    def _workload_initialise(self):
+        """Workload initialisation commands.
+
+        Create users and configuration to setup instance as an Group Replication node.
+        Raised errors must be treated on handlers.
+        """
+        self._mysql.configure_mysql_users()
+        self._mysql.configure_instance()
+        self._mysql.wait_until_mysql_connection()
+        workload_version = self._mysql.get_mysql_version()
+        self.unit.set_workload_version(workload_version)
+        self.unit_peer_data["instance-hostname"] = f"{instance_hostname()}:3306"
+
+    def _workload_reset(self, force: bool = False) -> StatusBase:
+        """Reset an errored workload.
+
+        Purge all files and re-initialise the workload.
+
+        Args:
+            force: flag to indicate if data needs to be forcefully purged
+
+        Returns:
+            A `StatusBase` to be set by the caller
+        """
+        # checks for data directory integrity
+        if not force and self._mysql.is_data_dir_initialised():
+            # avoid running for an apparently integral data dir
+            logger.error("Won't reset integral data directory.")
+            return BlockedStatus("data directory not empty")
+
+        try:
+            service_stop(SERVICE_NAME)
+            self._mysql.reset_data_dir()
+            self._mysql.reconfigure_mysqld()
+            self._workload_initialise()
+            unit_label = self.unit.name.replace("/", "-")
+            primary_address = self._get_primary_address_from_peers()
+            # On a full reset, member must firstly be removed from cluster metadata
+            self._mysql.remove_obsoletes_instance(from_instance=primary_address)
+            # Re-add the member as if it's the first time
+            self._mysql.add_instance_to_cluster(
+                self._mysql.instance_address, unit_label, from_instance=primary_address
+            )
+        except MySQLReconfigureError:
+            return BlockedStatus("Failed to re-initialize MySQL data-dir")
+        except MySQLConfigureMySQLUsersError:
+            return BlockedStatus("Failed to re-initialize MySQL users")
+        except MySQLConfigureInstanceError:
+            return BlockedStatus("Failed to re-configure instance for InnoDB")
+
+        return ActiveStatus()
+
+    def _get_primary_address_from_peers(self) -> str:
+        """Retrieve primary address based on peer data."""
+        for unit in self._peers.units:
+            if self._peers.data[unit]["member-role"] == "primary":
+                return self._peers.data[unit]["instance-hostname"]
```

Comment:
this may be stale until the next time the `update_status` hook runs. any reason why we dont just call `mysql.get_cluster_primary_address`?

Reply:
This is a workaround when the unit is in error, it cannot retrieve the primary by itself, so we resort to this alternative method to retrieve the primary.